Linkage Checker to work with artifacts with classifiers

Linkage Checker to work with artifacts with classifiers

https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/2162#discussion_r680151651

Given build.gradle:

```
plugins {
    id 'java'
    id 'com.google.cloud.tools.linkagechecker' version "1.5.9"
}

group 'org.example'
version '1.0-SNAPSHOT'

repositories {
    mavenCentral()
}

dependencies {
    compile 'com.google.cloud:google-cloud-core:1.95.4'
    compile 'com.google.cloud:google-cloud-core:1.95.4:tests'
}

test {
    useJUnitPlatform()
}

linkageChecker {
    configurations = ['compile']
}
```

Running the command below fails:

```
./gradlew linkageCheck --stacktrace
...
* What went wrong:
Execution failed for task ':linkageCheck'.
> com.google.cloud:google-cloud-core:1.95.4 is not in the class path
...
Caused by: java.lang.IllegalArgumentException: com.google.cloud:google-cloud-core:1.95.4 is not in the class path
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:219)
        at com.google.cloud.tools.opensource.classpath.ClassPathResult.formatDependencyPaths(ClassPathResult.java:68)
        at com.google.cloud.tools.opensource.classpath.LinkageProblem.dependencyPathsOfProblematicJars(LinkageProblem.java:248)
        at com.google.cloud.tools.opensource.classpath.LinkageProblem.formatLinkageProblems(LinkageProblem.java:228)
        at com.google.cloud.tools.dependencies.gradle.LinkageCheckTask.findLinkageErrors(LinkageCheckTask.java:159)
        at com.google.cloud.tools.dependencies.gradle.LinkageCheckTask.run(LinkageCheckTask.java:98)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)

```

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java
@@ -83,12 +83,13 @@ public final class ClassPathEntry {
   @Override
   public String toString() {
     if (artifact != null) {
-      if (!artifact.getClassifier().isEmpty()) {
-        return artifact.toString();
-      } else {
+      if (artifact.getClassifier().isEmpty()) {
         // If classifier is empty, this shows the shorter format of Maven coordinates without "jar"
         // extension
         return Artifacts.toCoordinates(artifact);
+      } else {
+        // This includes classifier
+        return artifact.toString();
       }
     } else {
       return jar.toString();

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java
@@ -83,9 +83,13 @@ public final class ClassPathEntry {
   @Override
   public String toString() {
     if (artifact != null) {
-      // Group ID, artifact ID and version. No extension such as "jar" or "tar.gz", because Linkage
-      // Checker uses only JAR artifacts.
-      return Artifacts.toCoordinates(artifact);
+      if (!artifact.getClassifier().isEmpty()) {
+        return artifact.toString();
+      } else {
+        // If classifier is empty, this shows the shorter format of Maven coordinates without "jar"
+        // extension
+        return Artifacts.toCoordinates(artifact);
+      }
     } else {
       return jar.toString();
     }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MavenDependencyMediation.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MavenDependencyMediation.java
@@ -34,6 +34,8 @@ class MavenDependencyMediation implements DependencyMediation {
   @Override
   public AnnotatedClassPath mediate(DependencyGraph dependencyGraph) {
     Set<Artifact> artifacts = new HashSet<>();
+
+    // Versionless coordinates plus classifier, if any
     Set<String> alreadyFound = new HashSet<>();
 
     AnnotatedClassPath annotatedClassPath = new AnnotatedClassPath();
@@ -44,7 +46,7 @@ class MavenDependencyMediation implements DependencyMediation {
 
       File file = artifact.getFile();
       if (file != null && file.getName().endsWith(".jar")) {
-        String versionlessCoordinates = Artifacts.makeKey(artifact);
+        String versionlessCoordinates = Artifacts.makeKey(artifact) + artifact.getClassifier();
         if (alreadyFound.add(versionlessCoordinates)) {
           // Adds to artifacts when versionlessCoordinates are new.
           artifacts.add(artifact);

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MavenDependencyMediation.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MavenDependencyMediation.java
@@ -46,7 +46,8 @@ class MavenDependencyMediation implements DependencyMediation {
 
       File file = artifact.getFile();
       if (file != null && file.getName().endsWith(".jar")) {
-        String versionlessCoordinates = Artifacts.makeKey(artifact) + artifact.getClassifier();
+        String versionlessCoordinates =
+            Artifacts.makeKey(artifact) + ":" + artifact.getClassifier();
         if (alreadyFound.add(versionlessCoordinates)) {
           // Adds to artifacts when versionlessCoordinates are new.
           artifacts.add(artifact);

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilderTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilderTest.java
@@ -25,6 +25,7 @@ import com.google.cloud.tools.opensource.dependencies.Bom;
 import com.google.cloud.tools.opensource.dependencies.UnresolvableArtifactProblem;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.truth.Correspondence;
 import com.google.common.truth.Truth;
 import com.google.common.truth.Truth8;
 import java.io.IOException;
@@ -281,5 +282,20 @@ public class ClassPathBuilderTest {
         classPathBuilder.resolve(
             ImmutableList.of(beamZetaSqlExtensions), true, DependencyMediation.MAVEN);
     assertNotNull(result);
+  }
+
+  @Test
+  public void testArtifactsWithClassifiers() throws Exception {
+    String pomFile = "bom-with-classifier-artifacts.pom";
+    Bom bom = Bom.readBom(TestHelper.absolutePathOfResource(pomFile));
+    ClassPathResult classPathResult =
+        classPathBuilder.resolve(bom.getManagedDependencies(), true, DependencyMediation.MAVEN);
+
+    Truth.assertThat(classPathResult.getClassPath())
+        .comparingElementsUsing(
+            Correspondence.transforming(
+                (ClassPathEntry entry) -> entry.getArtifact().toString(),
+                "has artifact coordinates"))
+        .contains("com.google.cloud:google-cloud-core:jar:tests:1.96.0");
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
@@ -37,6 +37,7 @@ import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Lists;
 import com.google.common.truth.Correspondence;
+import com.google.common.truth.Truth;
 import com.google.common.truth.Truth8;
 import java.io.IOException;
 import java.net.URISyntaxException;
@@ -1284,5 +1285,37 @@ public class LinkageCheckerTest {
           "Could not resolve 2 dependencies. See the message above for details.",
           expected.getMessage());
     }
+  }
+
+  @Test
+  public void testBomHavingClassifierArtifacts() throws Exception {
+    String pomFile = "bom-with-classifier-artifacts.pom";
+    Bom bom = Bom.readBom(TestHelper.absolutePathOfResource(pomFile));
+
+    LinkageChecker linkageChecker = LinkageChecker.create(bom);
+    ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
+
+    // bom-with-classifier-artifacts contains com.google.cloud:goole-cloud-core:jar:tests:1.96.0.
+    // This tests-classifier artifact requires Truth and JUnit dependencies to work, but it's not
+    // declared in its pom.xml (The user of the class needs to declare them because the tests-
+    // classifier artifacts are intended to be used in tests.) Therefore the linkage errors to the
+    // classes in the missing dependencies are expected.
+    Truth.assertThat(linkageProblems)
+        .comparingElementsUsing(
+            Correspondence.transforming(
+                (LinkageProblem problem) -> problem.getSymbol().getClassBinaryName(),
+                "has missing symbol"))
+        .containsAtLeast(
+            "com.google.common.testing.EqualsTester",
+            "com.google.common.truth.Truth",
+            "org.junit.Assert");
+
+    Truth.assertThat(linkageProblems)
+        .comparingElementsUsing(
+            Correspondence.transforming(
+                (LinkageProblem problem) ->
+                    problem.getSourceClass().getClassPathEntry().getArtifact().toString(),
+                "has source class belonging to artifact"))
+        .contains("com.google.cloud:google-cloud-core:jar:tests:1.96.0");
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
@@ -1295,7 +1295,7 @@ public class LinkageCheckerTest {
     LinkageChecker linkageChecker = LinkageChecker.create(bom);
     ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
 
-    // bom-with-classifier-artifacts contains com.google.cloud:goole-cloud-core:jar:tests:1.96.0.
+    // bom-with-classifier-artifacts contains com.google.cloud:google-cloud-core:jar:tests:1.96.0.
     // This tests-classifier artifact requires Truth and JUnit dependencies to work, but it's not
     // declared in its pom.xml (The user of the class needs to declare them because the tests-
     // classifier artifacts are intended to be used in tests.) Therefore the linkage errors to the

--- a/dependencies/src/test/resources/bom-with-classifier-artifacts.pom
+++ b/dependencies/src/test/resources/bom-with-classifier-artifacts.pom
@@ -1,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2021 Google LLC.
+  ~
+  ~ Licensed under the Apache License, Version 2.0 (the "License");
+  ~ you may not use this file except in compliance with the License.
+  ~ You may obtain a copy of the License at
+  ~
+  ~     http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<!-- This file is copied from com.google.cloud:libraries-bom:2.7.0 for RepositoryUtilityTest -->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0
+                             http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.google.cloud</groupId>
+  <artifactId>bom-with-classifier-artifacts</artifactId>
+  <version>0.0.1</version>
+  <packaging>pom</packaging>
+
+  <name>BOM that contains artifacts with classifiers</name>
+  <url>https://github.com/GoogleCloudPlatform/cloud-opensource-java#google-library-bom</url>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  </properties>
+
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>com.google.cloud</groupId>
+        <artifactId>google-cloud-core</artifactId>
+        <version>1.96.0</version>
+      </dependency>
+      <dependency>
+        <groupId>com.google.cloud</groupId>
+        <artifactId>google-cloud-core</artifactId>
+        <version>1.96.0</version>
+        <!--
+          This test-jar type points to the same JAR file as tests classifier
+          https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/2162#discussion_r680193168
+         -->
+        <type>test-jar</type>
+      </dependency>
+      <dependency>
+        <groupId>com.google.cloud</groupId>
+        <artifactId>google-cloud-core</artifactId>
+        <version>1.96.0</version>
+        <classifier>tests</classifier>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
+</project>

--- a/dependencies/src/test/resources/bom-with-classifier-artifacts.pom
+++ b/dependencies/src/test/resources/bom-with-classifier-artifacts.pom
@@ -15,7 +15,6 @@
   ~ limitations under the License.
   -->
 
-<!-- This file is copied from com.google.cloud:libraries-bom:2.7.0 for RepositoryUtilityTest -->
 <project xmlns="http://maven.apache.org/POM/4.0.0"
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://maven.apache.org/POM/4.0.0

--- a/gradle-plugin/src/functionalTest/groovy/com/google/cloud/tools/dependencies/gradle/BuildStatusFunctionalTest.groovy
+++ b/gradle-plugin/src/functionalTest/groovy/com/google/cloud/tools/dependencies/gradle/BuildStatusFunctionalTest.groovy
@@ -113,4 +113,39 @@ class BuildStatusFunctionalTest extends Specification {
         |""".stripMargin())
     result.task(":linkageCheck").outcome == TaskOutcome.FAILED
   }
+
+  def "can handle artifacts with classifiers "() {
+    buildFile << """
+        repositories {
+          mavenCentral()
+        }
+        
+        dependencies {
+          // The tests-classifier artifacts should not stop Linkage Checker
+          compile 'com.google.cloud:google-cloud-core:1.95.4'
+          compile 'com.google.cloud:google-cloud-core:1.95.4:tests'
+        }
+        
+        linkageChecker {
+          configurations = ['compile']
+        }
+        """
+
+    when:
+    def result = GradleRunner.create()
+        .withProjectDir(testProjectDir.root)
+        .withArguments('linkageCheck', '--stacktrace')
+        .withPluginClasspath()
+        .buildAndFail()
+
+    then:
+    result.output.contains("Task :linkageCheck")
+    result.output.contains("Linkage Checker rule found 72 errors:")
+    result.output.contains("""Class org.junit.Assert is not found;
+        |  referenced by 26 class files
+        |    com.google.cloud.ServiceOptionsTest (com.google.cloud:google-cloud-core:jar:tests:1.95.4)
+        |    com.google.cloud.BatchResultTest (com.google.cloud:google-cloud-core:jar:tests:1.95.4)
+        |  """.stripMargin())
+    result.task(":linkageCheck").outcome == TaskOutcome.FAILED
+  }
 }

--- a/gradle-plugin/src/functionalTest/groovy/com/google/cloud/tools/dependencies/gradle/BuildStatusFunctionalTest.groovy
+++ b/gradle-plugin/src/functionalTest/groovy/com/google/cloud/tools/dependencies/gradle/BuildStatusFunctionalTest.groovy
@@ -148,4 +148,39 @@ class BuildStatusFunctionalTest extends Specification {
         |  """.stripMargin())
     result.task(":linkageCheck").outcome == TaskOutcome.FAILED
   }
+
+  def "can handle artifacts with classifiers with transitive dependencies"() {
+    buildFile << """
+        repositories {
+          mavenCentral()
+        }
+        
+        dependencies {
+          // google-cloud-bigquery depends on google-cloud-core (no classifier)
+          compile 'com.google.cloud:google-cloud-bigquery:1.137.1'
+          compile 'com.google.cloud:google-cloud-core:1.95.4:tests'
+        }
+        
+        linkageChecker {
+          configurations = ['compile']
+        }
+        """
+
+    when:
+    def result = GradleRunner.create()
+        .withProjectDir(testProjectDir.root)
+        .withArguments('linkageCheck', '--stacktrace')
+        .withPluginClasspath()
+        .buildAndFail()
+
+    then:
+    result.output.contains("Task :linkageCheck")
+    result.output.contains("Linkage Checker rule found 93 errors")
+    result.output.contains("""Class org.junit.Assert is not found;
+        |  referenced by 26 class files
+        |    com.google.cloud.ServiceOptionsTest (com.google.cloud:google-cloud-core:jar:tests:1.95.4)
+        |    com.google.cloud.BatchResultTest (com.google.cloud:google-cloud-core:jar:tests:1.95.4)
+        |  """.stripMargin())
+    result.task(":linkageCheck").outcome == TaskOutcome.FAILED
+  }
 }

--- a/gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java
+++ b/gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java
@@ -30,11 +30,9 @@ import com.google.cloud.tools.opensource.dependencies.Artifacts;
 import com.google.cloud.tools.opensource.dependencies.DependencyGraph;
 import com.google.cloud.tools.opensource.dependencies.DependencyPath;
 import com.google.cloud.tools.opensource.dependencies.PathToNode;
-import com.google.common.base.Verify;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableListMultimap;
 import com.google.common.collect.ImmutableSet;
-import java.io.File;
 import java.io.IOException;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -285,7 +283,7 @@ public class LinkageCheckTask extends DefaultTask {
     return new Dependency(artifact, "compile");
   }
 
-  private static DependencyGraph createDependencyGraph(ResolvedConfiguration configuration) {
+  private DependencyGraph createDependencyGraph(ResolvedConfiguration configuration) {
     // Why this method is not part of the DependencyGraph? Because the dependencies module
     // which the DependencyGraph belongs to is a Maven project, and Gradle does not provide good
     // Maven artifacts to develop code with Gradle-related classes.
@@ -311,6 +309,14 @@ public class LinkageCheckTask extends DefaultTask {
 
       DependencyPath path = null;
 
+      if (node.getModuleArtifacts().isEmpty()) {
+        getLogger()
+            .warn(
+                "The dependency node "
+                    + node.getName()
+                    + " should have at least one resolved artifacts");
+        continue;
+      }
       // For artifacts with classifiers, there can be multiple resolved artifacts for one node
       for (ResolvedArtifact artifact : node.getModuleArtifacts()) {
         // parentPath is null for the first item
@@ -321,8 +327,6 @@ public class LinkageCheckTask extends DefaultTask {
         graph.addPath(path);
       }
 
-      Verify.verify(
-          path != null, "The dependency node should have at least one resolved artifacts");
       for (ResolvedDependency child : node.getChildren()) {
         if (visited.add(child)) {
           queue.add(new PathToNode<>(child, path));
@@ -333,7 +337,7 @@ public class LinkageCheckTask extends DefaultTask {
     return graph;
   }
 
-  private static ClassPathResult createClassPathResult(ResolvedConfiguration configuration) {
+  private ClassPathResult createClassPathResult(ResolvedConfiguration configuration) {
     DependencyGraph dependencyGraph = createDependencyGraph(configuration);
     AnnotatedClassPath annotatedClassPath = new AnnotatedClassPath();
 

--- a/gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java
+++ b/gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java
@@ -30,6 +30,7 @@ import com.google.cloud.tools.opensource.dependencies.Artifacts;
 import com.google.cloud.tools.opensource.dependencies.DependencyGraph;
 import com.google.cloud.tools.opensource.dependencies.DependencyPath;
 import com.google.cloud.tools.opensource.dependencies.PathToNode;
+import com.google.common.base.Verify;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableListMultimap;
 import com.google.common.collect.ImmutableSet;
@@ -120,8 +121,8 @@ public class LinkageCheckTask extends DefaultTask {
           new DefaultArtifact(
               moduleVersionId.getGroup(),
               moduleVersionId.getName(),
-              null,
-              null,
+              resolvedArtifact.getClassifier(),
+              resolvedArtifact.getExtension(),
               moduleVersionId.getVersion(),
               null,
               resolvedArtifact.getFile());
@@ -263,6 +264,21 @@ public class LinkageCheckTask extends DefaultTask {
     return output.toString();
   }
 
+  private static Artifact artifactFrom(
+      ResolvedDependency resolvedDependency, ResolvedArtifact resolvedArtifact) {
+    ModuleVersionIdentifier moduleVersionId = resolvedDependency.getModule().getId();
+    DefaultArtifact artifact =
+        new DefaultArtifact(
+            moduleVersionId.getGroup(),
+            moduleVersionId.getName(),
+            resolvedArtifact.getClassifier(),
+            resolvedArtifact.getExtension(),
+            moduleVersionId.getVersion(),
+            null,
+            resolvedArtifact.getFile());
+    return artifact;
+  }
+
   private static Artifact artifactFrom(ResolvedDependency resolvedDependency) {
     ModuleVersionIdentifier moduleVersionId = resolvedDependency.getModule().getId();
     File file = resolvedDependency.getModuleArtifacts().iterator().next().getFile();
@@ -278,8 +294,9 @@ public class LinkageCheckTask extends DefaultTask {
     return artifact;
   }
 
-  private static Dependency dependencyFrom(ResolvedDependency resolvedDependency) {
-    Artifact artifact = artifactFrom(resolvedDependency);
+  private static Dependency dependencyFrom(
+      ResolvedDependency resolvedDependency, ResolvedArtifact resolvedArtifact) {
+    Artifact artifact = artifactFrom(resolvedDependency, resolvedArtifact);
     return new Dependency(artifact, "compile");
   }
 
@@ -307,14 +324,20 @@ public class LinkageCheckTask extends DefaultTask {
 
       DependencyPath parentPath = item.getParentPath();
 
-      // parentPath is null for the first item
-      DependencyPath path =
-          parentPath == null
-              ? new DependencyPath(artifactFrom(node))
-              : parentPath.append(dependencyFrom(node));
+      DependencyPath path = null;
 
-      graph.addPath(path);
+      // When there are artifacts with classifiers, there are multiple resolved artifacts
+      for (ResolvedArtifact artifact : node.getModuleArtifacts()) {
+        // parentPath is null for the first item
+        path =
+            parentPath == null
+                ? new DependencyPath(artifactFrom(node, artifact))
+                : parentPath.append(dependencyFrom(node, artifact));
+        graph.addPath(path);
+      }
 
+      Verify.verify(
+          path != null, "The dependency node should have at least one resolved artifacts");
       for (ResolvedDependency child : node.getChildren()) {
         if (visited.add(child)) {
           queue.add(new PathToNode<>(child, path));
@@ -324,6 +347,7 @@ public class LinkageCheckTask extends DefaultTask {
 
     return graph;
   }
+
 
   private static ClassPathResult createClassPathResult(ResolvedConfiguration configuration) {
     DependencyGraph dependencyGraph = createDependencyGraph(configuration);

--- a/gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java
+++ b/gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java
@@ -307,9 +307,12 @@ public class LinkageCheckTask extends DefaultTask {
 
       DependencyPath parentPath = item.getParentPath();
 
+      // If there are multiple artifacts (with different classifiers) in this node, then the path is
+      // the same, because these artifacts share the same dependencies with the same pom.xml.
       DependencyPath path = null;
 
-      if (node.getModuleArtifacts().isEmpty()) {
+      Set<ResolvedArtifact> moduleArtifacts = node.getModuleArtifacts();
+      if (moduleArtifacts.isEmpty()) {
         getLogger()
             .warn(
                 "The dependency node "
@@ -317,8 +320,9 @@ public class LinkageCheckTask extends DefaultTask {
                     + " should have at least one resolved artifacts");
         continue;
       }
+
       // For artifacts with classifiers, there can be multiple resolved artifacts for one node
-      for (ResolvedArtifact artifact : node.getModuleArtifacts()) {
+      for (ResolvedArtifact artifact : moduleArtifacts) {
         // parentPath is null for the first item
         path =
             parentPath == null

--- a/gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java
+++ b/gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java
@@ -279,21 +279,6 @@ public class LinkageCheckTask extends DefaultTask {
     return artifact;
   }
 
-  private static Artifact artifactFrom(ResolvedDependency resolvedDependency) {
-    ModuleVersionIdentifier moduleVersionId = resolvedDependency.getModule().getId();
-    File file = resolvedDependency.getModuleArtifacts().iterator().next().getFile();
-    DefaultArtifact artifact =
-        new DefaultArtifact(
-            moduleVersionId.getGroup(),
-            moduleVersionId.getName(),
-            null,
-            null,
-            moduleVersionId.getVersion(),
-            null,
-            file);
-    return artifact;
-  }
-
   private static Dependency dependencyFrom(
       ResolvedDependency resolvedDependency, ResolvedArtifact resolvedArtifact) {
     Artifact artifact = artifactFrom(resolvedDependency, resolvedArtifact);
@@ -326,7 +311,7 @@ public class LinkageCheckTask extends DefaultTask {
 
       DependencyPath path = null;
 
-      // When there are artifacts with classifiers, there are multiple resolved artifacts
+      // For artifacts with classifiers, there can be multiple resolved artifacts for one node
       for (ResolvedArtifact artifact : node.getModuleArtifacts()) {
         // parentPath is null for the first item
         path =
@@ -347,7 +332,6 @@ public class LinkageCheckTask extends DefaultTask {
 
     return graph;
   }
-
 
   private static ClassPathResult createClassPathResult(ResolvedConfiguration configuration) {
     DependencyGraph dependencyGraph = createDependencyGraph(configuration);

